src/progress/progress_set.rs

```diff
@@ -43,91 +44,98 @@ pub fn majority(total: usize) -> usize {
 /// A Raft internal representation of a Configuration.
 ///
 /// This is corollary to a ConfState, but optimized for `contains` calls.
-#[derive(Clone, Debug, Default, PartialEq, Getters)]
+#[derive(Clone, Debug, PartialEq, Default)]
 pub struct Configuration {
-    /// The voter set.
-    #[get = "pub"]
-    voters: HashSet<u64>,
-    /// The learner set.
-    #[get = "pub"]
-    learners: HashSet<u64>,
-}
+    auto_leave: bool,
 
-impl Configuration {
-    /// Create a new configuration with the given configuration.
-    pub fn new(
-        voters: impl IntoIterator<Item = u64>,
-        learners: impl IntoIterator<Item = u64>,
-    ) -> Self {
-        Self {
-            voters: voters.into_iter().collect(),
-            learners: learners.into_iter().collect(),
-        }
-    }
+    // Sorted voters. Only the first is valid if it's not in joint.
+    // Otherwise the first is incoming and the second is outgoing.
+    voters: [Vec<u64>; 2],
 
-    /// Create a new `ConfState` from the configuration itself.
-    pub fn to_conf_state(&self) -> ConfState {
-        let mut state = ConfState::default();
-        state.set_voters(self.voters.iter().cloned().collect());
-        state.set_learners(self.learners.iter().cloned().collect());
-        state
-    }
+    // Sorted learners. Shouldn't intersect with `voters`.
+    learners: Vec<u64>,
 
-    /// Create a new `Configuration` from a given `ConfState`.
-    pub fn from_conf_state(conf_state: &ConfState) -> Self {
-        Self {
-            voters: conf_state.voters.iter().cloned().collect(),
-            learners: conf_state.learners.iter().cloned().collect(),
-        }
+    // Demoted learners in joint consensus, because `learners` shouldn't intersect with `voters`.
+    learners_next: Vec<u64>,
```

Comment:
In `voters` we use `[Vec<u64>; 2]`, can we not do this with `learners`?

Reply:
Of course we can. However, etcd/raft uses `learners_next`, so I guess it's better to keep it.